spring-integration-core/src/main/java/org/springframework/integration/handler/SuspendingMessageHandler.java

```diff
@@ -0,0 +1,132 @@
+
+import org.springframework.integration.IntegrationMessageHeaderAccessor;
+import org.springframework.integration.aggregator.CorrelationStrategy;
+import org.springframework.integration.aggregator.HeaderAttributeCorrelationStrategy;
+import org.springframework.messaging.Message;
+import org.springframework.messaging.MessagingException;
+
+/**
+ * A pair of message handlers; the first suspends the thread until the second receives
+ * a corresponding message.
+ *
+ * @author Gary Russell
+ * @since 4.2
+ *
+ */
+public class SuspendingMessageHandler extends AbstractReplyProducingMessageHandler {
+
+	private final ConcurrentHashMap<Object, BlockingQueue<Message<?>>> suspensions =
+			new ConcurrentHashMap<Object, BlockingQueue<Message<?>>>();
+
+	private final ConcurrentHashMap<Object, Thread> inProcess = new ConcurrentHashMap<Object, Thread>();
+
+	private final long timeout;
+
+	private final CorrelationStrategy correlationStrategy;
+
+	public SuspendingMessageHandler(long timeout) {
+		this(new HeaderAttributeCorrelationStrategy(IntegrationMessageHeaderAccessor.CORRELATION_ID), timeout);
+	}
+
+	public SuspendingMessageHandler(CorrelationStrategy correlationStrategy,
+			long timeout) {
+		this.correlationStrategy = correlationStrategy;
+		this.timeout = timeout;
+	}
+
+	@Override
+	protected Object handleRequestMessage(Message<?> requestMessage) {
+		Object key = SuspendingMessageHandler.this.correlationStrategy.getCorrelationKey(requestMessage);
+		if (key == null) {
+			throw new MessagingException(requestMessage, "Correlation Strategy returned null");
+		}
+		Thread existing = this.inProcess.putIfAbsent(key, Thread.currentThread());
+		if (existing != null) {
+			throw new MessagingException(requestMessage, "Correlation key ("
+					+ key + ") is already in use by " + existing.getName());
+		}
+		BlockingQueue<Message<?>> blockingQueue = createOrObtainQueue(key);
+		try {
+			Message<?> release = blockingQueue.poll(this.timeout, TimeUnit.MILLISECONDS);
+			if (release != null) {
+				this.suspensions.remove(key);
+				if (release.getPayload() instanceof Throwable) {
+					throw new MessagingException(requestMessage, "Releasing flow returned a throwable",
+							(Throwable) release.getPayload());
+				}
+				else {
+					return buildResult(requestMessage, release);
+				}
+			}
+			else {
+				synchronized(blockingQueue) {
+					if (!blockingQueue.offer(requestMessage)) {
+						// second chance
+						return buildResult(requestMessage, blockingQueue.poll());
+					}
+				}
+				return null;
+			}
+		}
+		catch (InterruptedException e) {
+			Thread.currentThread().interrupt();
+			throw new MessagingException(requestMessage, "Interrupted while waiting for release", e);
+		}
+		finally {
+			this.inProcess.remove(key);
+		}
+	}
+
+	private Object buildResult(Message<?> requestMessage, Message<?> release) {
+		Object[] result = new Object[2];
+		result[0] = requestMessage.getPayload();
+		result[1] = release;
+		return result;
```

Comment:
I think we must provide the hook to inject `MessageListProcessor` to allow to decide what to do both those messages. 
By default we can do the same what `AggregatingMessageHandler` does - `DefaultAggregatingMessageGroupProcessor`.